File: python/intervels/MeetingRoomsII.py
'''
Given an array of meeting time interval objects consisting of start and end times [[start_1,end_1],[start_2,end_2],...] (start_i < end_i), find the minimum number of days required to schedule all meetings without any conflicts.

Note: (0,8),(8,10) is not considered a conflict at 8.

Example 1:
Input: intervals = [(0,40),(5,10),(15,20)]
Output: 2
Explanation:
day1: (0,40)
day2: (5,10),(15,20)

Example 2:
Input: intervals = [(4,9)]
Output: 1
'''
from typing import List

from Interval import Interval

class MeetingRoomsII:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if not intervals:
            return 0
            
        # Separate and sort start and end times
        starts = sorted([i.start for i in intervals])
        ends = sorted([i.end for i in intervals])
        
        res, count = 0, 0
        s, e = 0, 0
        
        # Iterate through all start times
        while s < len(intervals):
            # If a meeting starts before the earliest ending meeting finishes, we need a new room
            if starts[s] < ends[e]:
                count += 1
                s += 1
            # A meeting ended, free up a room
            else:
                count -= 1
                e += 1
            res = max(res, count)
            
        return res

if __name__ == "__main__":
    solution = MeetingRoomsII()
    
    # Test Case 1
    intervals1 = [Interval(0, 40), Interval(5, 10), Interval(15, 20)]
    print(f"Input: {intervals1}")
    print(f"Output: {solution.minMeetingRooms(intervals1)}") # Expected: 2
    print("-" * 20)

    # Test Case 2
    intervals2 = [Interval(4, 9)]
    print(f"Input: {intervals2}")
    print(f"Output: {solution.minMeetingRooms(intervals2)}") # Expected: 1
    print("-" * 20)
    
    # Test Case 3 (Touching intervals)
    intervals3 = [Interval(0, 8), Interval(8, 10)]
    print(f"Input: {intervals3}")
    print(f"Output: {solution.minMeetingRooms(intervals3)}") # Expected: 1

File: src/main/java/com/naresh/a_dsalgo/Intervals/problems/MeetingRoomsII.java
package com.naresh.a_dsalgo.Intervals.problems;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Problem: Meeting Rooms II (LeetCode 253)
 * Description: Given an array of meeting time intervals `intervals` where `intervals[i] = [starti, endi]`,
 * return the minimum number of conference rooms required.
 */
public class MeetingRoomsII {

    /**
     * Algorithm: Greedy approach using a min-heap.
     * 1. Sort all intervals by their start times.
     * 2. Initialize a min-heap to store the end times of meetings. This heap will always give us
     *    the meeting that finishes earliest among those currently occupying rooms.
     * 3. Iterate through the sorted intervals:
     *    a. If the heap is not empty and the current meeting's start time is greater than or equal to
     *       the earliest ending meeting in the heap (`minHeap.peek()`), it means a room becomes free.
     *       Remove that meeting's end time from the heap (`minHeap.poll()`).
     *    b. Add the current meeting's end time to the heap (`minHeap.offer(interval[1])`). This meeting
     *       now occupies a room.
     * 4. The maximum size the heap reaches during the iteration is the minimum number of rooms required.
     *    The final size of the heap will be this maximum.
     *
     * @param intervals A 2D array of meeting time intervals `[start, end]`.
     * @return The minimum number of conference rooms required.
     */
    public int minMeetingRooms(int[][] intervals) {
        // Pattern: Greedy (Sorting + Min-Heap) | Time: O(N log N), Space: O(N)
        if (intervals == null || intervals.length == 0) return 0;

        // Sort intervals by their start times
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));

        // Min-heap to store the end times of meetings currently in progress
        var minHeap = new PriorityQueue<Integer>();
        minHeap.offer(intervals[0][1]); // Add the end time of the first meeting

        for (var i = 1; i < intervals.length; i++) {
            // If the current meeting's start time is >= the earliest ending meeting,
            // it means a room is free.
            if (intervals[i][0] >= minHeap.peek()) {
                minHeap.poll(); // Free up that room
            }
            minHeap.offer(intervals[i][1]); // Occupy a room with the current meeting's end time
        }
        return minHeap.size(); // The number of rooms needed is the max size of the heap
    }
    // FAANG Tip: This problem is a classic application of a min-heap to track resource allocation (meeting rooms). Sorting by start time is crucial.

    public static void main(String[] args) {
        var solution = new MeetingRoomsII();

        // Test Case 1: Example from LeetCode
        var intervals1 = new int[][]{{0, 30}, {5, 10}, {15, 20}};
        // Expected: 2 (e.g., [0,30] in Room 1, [5,10] in Room 2, [15,20] in Room 2 after [5,10] ends)
        System.out.println("Intervals: " + Arrays.deepToString(intervals1) + " -> Min Rooms: " + solution.minMeetingRooms(intervals1));

        // Test Case 2: Example from LeetCode
        var intervals2 = new int[][]{{7, 10}, {2, 4}};
        // Expected: 1
        System.out.println("Intervals: " + Arrays.deepToString(intervals2) + " -> Min Rooms: " + solution.minMeetingRooms(intervals2));

        // Test Case 3: No overlap
        var intervals3 = new int[][]{{1, 2}, {3, 4}, {5, 6}};
        // Expected: 1
        System.out.println("Intervals: " + Arrays.deepToString(intervals3) + " -> Min Rooms: " + solution.minMeetingRooms(intervals3));

        // Test Case 4: All overlap
        var intervals4 = new int[][]{{1, 10}, {2, 9}, {3, 8}};
        // Expected: 3
        System.out.println("Intervals: " + Arrays.deepToString(intervals4) + " -> Min Rooms: " + solution.minMeetingRooms(intervals4));

        // Test Case 5: Empty intervals
        var intervals5 = new int[][]{};
        // Expected: 0
        System.out.println("Intervals: " + Arrays.deepToString(intervals5) + " -> Min Rooms: " + solution.minMeetingRooms(intervals5));

        // Test Case 6: Single interval
        var intervals6 = new int[][]{{1, 10}};
        // Expected: 1
        System.out.println("Intervals: " + Arrays.deepToString(intervals6) + " -> Min Rooms: " + solution.minMeetingRooms(intervals6));
    }
}
